src/bp/nlu-server/api.ts

```diff
@@ -0,0 +1,237 @@
+    debugRequest('incoming ' + req.path, { ip: req.ip })
+    next()
+  })
+
+  app.use(monitoringMiddleware)
+  app.use(handleUnexpectedError)
+
+  if (process.core_env.REVERSE_PROXY) {
+    app.set('trust proxy', process.core_env.REVERSE_PROXY)
+  }
+
+  if (options.limit > 0) {
+    app.use(
+      rateLimit({
+        windowMs: ms(options.limitWindow),
+        max: options.limit,
+        message: 'Too many requests, please slow down'
+      })
+    )
+  }
+
+  if (options.authToken?.length) {
+    app.use(authMiddleware(options.authToken))
+  }
+
+  return app
+}
+
+export default async function(options: APIOptions) {
+  const app = createExpressApp(options)
+  const logger = new NLUServerLogger('API')
+  const loggerWrapper: NLU.Logger = {
+    info: (msg: string) => logger.info(msg),
+    warning: (msg: string, err?: Error) => (err ? logger.attachError(err).warn(msg) : logger.warn(msg)),
+    error: (msg: string, err?: Error) => (err ? logger.attachError(err).error(msg) : logger.error(msg))
+  }
+
+  let engine: Engine
+  let modelService: ModelService
+  let trainSessionService: TrainSessionService
+  let trainService: TrainService
+  try {
+    engine = new Engine('nlu-server', loggerWrapper)
+    modelService = new ModelService(options.modelDir)
+    await modelService.init()
+    trainSessionService = new TrainSessionService()
+    trainService = new TrainService(logger, engine, modelService, trainSessionService)
+  } catch (err) {
+    logger.attachError(err).error('an error occured while initializing the server')
+    process.exit(1)
+  }
+
+  app.get('/info', (req, res) => {
+    res.send({
+      version: options.version
+    })
+  })
+
+  const router = express.Router({ mergeParams: true })
+  router.post('/train', async (req, res) => {
+    try {
+      const input: TrainInput = await validate(req.body, TrainInputCreateSchema, {
+        stripUnknown: true
+      })
+
+      const { topics, entities, language, password } = input
+      const intents = _.flatMap(Object.values(topics))
+      const modelHash = engine.computeModelHash(intents, input.entities, input.language)
+
+      const seed = input.seed ?? Math.round(Math.random() * 10000)
+      const modelId = modelService.makeModelId(modelHash, input.language, seed)
+      const modelFileName = modelService.makeFileName(modelId, password)
+
+      // return the modelId as fast as possible
+      // tslint:disable-next-line: no-floating-promises
+      trainService.train(modelFileName, intents, entities, language, seed)
+
+      return res.send({
+        success: true,
+        modelId
+      })
+    } catch (err) {
+      res.status(500).send({
+        success: false,
+        error: err.message
+      })
+    }
+  })
+
+  router.get('/train/:modelId', async (req, res) => {
+    try {
+      const { modelId } = req.params
+      const { password } = req.query
+      const modelFileName = modelService.makeFileName(modelId, password ?? '')
```

Comment:
Why not make password optional instead ?

Reply:
I'm just getting fed up of optional typings. The faster I can resolve a value for an optional type, the cleaner the lower level code gets.

Reply:
Optional typings are there for a reason. It means that the value can be undefined. If you just keep making everything non optional, then we will just find even more bugs, which could have been avoided with the correct typings. 

If you do the right validations (or returning early), then you won't have to add ! everywhere. For ex:

```js
const makeFileName(modelId: string, password?: string){
  if(!password){
    return
  }
  // password is sure to be set
}
````

Reply:
No you're getting it all wrong.

This is a spec of the NLU-Server that an empty password should be considered as an empty string password. The HTTP API doesnt require a password, but the rest of the code does.

The password is actually mandatory to make a file name. The name is made by hashing the modelId with the password to create a uniq file name that can only be recovered with the correct modelId and the correct password. You could at least read the code of the `makeFileName` function before commeting my code and lecturing me on the purposes of typescript language features.

What I'm saying is not that optionnal type have no purposes, its that they should only be used when a value is truely optionnal and here its not.